x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -266,6 +266,7 @@ private static boolean shouldAuthorizeIndexActionNameOnly(String action, Transpo
             case "indices:data/write/reindex":
             case "indices:data/read/sql":
             case "indices:data/read/sql/translate":
+            case "indices:data/read/esql":
```

Comment:
@ywangd This is a big PR, and I do not expect a full review, but I do want to call out this change.

Reply:
Thanks for the ping. This particular change here looks right to me after briefly reading through relevant code around this change. Similar to the SQL API, the indices that ES|QL accesses is not known when the request is initially authorized. Therefore we only authorize it by the action names without looking into the indices. The index authorization will be performed when the ES|QL services send out shard level request for data access. These shard level requests are

1. [DataNodeRequest](https://github.com/elastic/elasticsearch/blob/2e4440ff1c0decf79f8a40688566411f7d518aa7/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequest.java#L32) - for data accessing and computation
2. [LookupRequest](https://github.com/elastic/elasticsearch/blob/2e4440ff1c0decf79f8a40688566411f7d518aa7/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/EnrichLookupService.java#L268) - for data encrichment

Both requests implement `IndicesRequest`, have corresponding transportHandlers, and are issued with `transportService.sendChildRequest`. Therefore they go through the existing index authorization process and properly authorized. 

_One question_ I do have is why we chose to [execute the `LookupRequest` with the `enrich` origin](https://github.com/elastic/elasticsearch/blob/2e4440ff1c0decf79f8a40688566411f7d518aa7/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/EnrichLookupService.java#L141C2-L141C2). It means the following query works when the user has access to only the `playlist` index without access to either the `songs` enrich policy or its associated index. 

```
FROM playlist | ENRICH songs ON song_id
```

It feels a bit weird to me that we have no control on who can use what enrich process. That said, we don't have a good story around enrich and ingest pipeline either, i.e. we have no control over who can use what pipeline where a pipeline can contain an enrich processor. There have been quite a few complaints from users and we have an [open issue](https://github.com/elastic/elasticsearch/issues/57586) for it. I wonder whether we should *not* repeat the problem here with ES|QL. On the surface, it might be easier to tackle compared to ingest pipelines since ES|QL is read-only and has no "bulk" operations. What do you think?

Reply:
Thank you so much @ywangd . This is a great question / suggestion, I filed the following issue to track it   #98482

Reply:
@ywangd The primary reason we perform the enrich lookup with the enrich origin is to prevent the warning that arises when accessing enrich indices, which are system indices. An alternative is to override `allowSystemIndexAccessByDefault` in the rest layer, but I felt that it might be too broad. I'm happy to iterate on it.

Reply:
I posted my reply on the issue (#98482) raised by Chris to have future discussions in one place and avoid disturbing all people involved in this huge PR.